finality-aleph/src/base_protocol/handler.rs

```diff
@@ -1,32 +1,89 @@
 
-use sc_network::{ExtendedPeerInfo, PeerId};
-use sc_network_common::role::Roles;
-use sp_runtime::traits::{Block, Header};
+use parity_scale_codec::DecodeAll;
+use sc_network::{config::FullNetworkConfiguration, ExtendedPeerInfo, PeerId};
+use sc_network_common::{role::Roles, sync::message::BlockAnnouncesHandshake};
+use sp_runtime::traits::{Block, Header, Saturating};
 
 use crate::{BlockHash, BlockNumber};
 
+pub enum ConnectError {
+    BadlyEncodedHandshake,
+    BadHandshakeGenesis,
+    PeerAlreadyConnected,
+    TooManyFullInboundPeers,
+    TooManyFullOutboundPeers,
+    TooManyLightPeers,
+}
+
+pub enum DisconnectError {
+    PeerWasNotConnected,
+}
+
 struct PeerInfo {
     role: Roles,
+    is_inbound: bool,
 }
 
 /// Handler for the base protocol.
-pub struct Handler {
+pub struct Handler<B>
+where
+    B: Block<Hash = BlockHash>,
+    B::Header: Header<Number = BlockNumber>,
+{
+    reserved_nodes: HashSet<PeerId>,
     peers: HashMap<PeerId, PeerInfo>,
+    // the below counters and bounds ignore the nodes which belong to `reserved_nodes`
+    num_full_in_peers: usize,
+    num_full_out_peers: usize,
+    num_light_peers: usize,
+    max_full_in_peers: usize,
+    max_full_out_peers: usize,
+    max_light_peers: usize,
+    genesis_hash: B::Hash,
 }
 
-impl Handler {
+impl<B> Handler<B>
+where
+    B: Block<Hash = BlockHash>,
+    B::Header: Header<Number = BlockNumber>,
+{
     /// Create a new handler.
-    pub fn new() -> Self {
+    pub fn new(genesis_hash: B::Hash, net_config: &FullNetworkConfiguration) -> Self {
+        let reserved_nodes = net_config
+            .network_config
+            .default_peers_set
+            .reserved_nodes
+            .iter()
+            .map(|reserved| reserved.peer_id)
+            .collect();
+
+        // It is assumed that `default_peers_set.out_peers` only refers to full nodes, but
+        // `default_peers_set.in_peers` refers to both full and light nodes.
+        // Moreover, `default_peers_set_num_full` refers to the total of full nodes.
+        let max_full_out_peers = net_config.network_config.default_peers_set.out_peers as usize;
+        let max_full_in_peers = (net_config.network_config.default_peers_set_num_full as usize)
+            .saturating_sub(max_full_out_peers);
+        let max_light_peers = (net_config.network_config.default_peers_set.in_peers as usize)
+            .saturating_sub(max_full_in_peers);
+
         Handler {
+            reserved_nodes,
             peers: HashMap::new(),
+            max_full_in_peers,
+            max_full_out_peers,
+            max_light_peers,
+            num_full_in_peers: 0,
+            num_full_out_peers: 0,
+            num_light_peers: 0,
+            genesis_hash,
         }
     }
 
     /// Returns a list of connected peers with some additional information.
     // TODO(A0-3886): This shouldn't need to return the substrate type after replacing RPCs.
     // In particular, it shouldn't depend on `B`.
```

Comment:
does this comment now relate to the whole struct? just asking

Reply:
It's actually more annoying than that, because we cannot really remove the `B` due to the handshake type. :/ However the comment here should be changed to at least remove the dependency on `ExtendedPeerInfo`.

If we ever want to remove the `B` completely I think we will have to add handshake versioning, I don't remember whether we have a task for that, but if we do it might be a good idea to link _that_ task by the whole struct.

An alternative would be to decode the handshake _before_ putting it in here – perhaps that's even a reasonable approach? Then, for now, both the methods can be parametrized by `B` or, even better, we can put a different, unparametrized, type into the verification/peer connected method.

Reply:
Removed the lowest line of the comment. We can see in further prs what we can do with removing dependencies.